atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -265,33 +305,29 @@ public ExpectationsAwareTransaction delegate() {
         }
 
         @Override
-        public <T, E extends Exception> T finish(TransactionTask<T, E> task)
-                throws E, TransactionFailedRetriableException {
-            return finishWithCallback(task, () -> {});
-        }
-
-        @Override
-        public <T, E extends Exception> T finishWithCallback(TransactionTask<T, E> task, Runnable callback)
-                throws E, TransactionFailedRetriableException {
-            Timer postTaskTimer = getTimer("finishTask");
-            Timer.Context postTaskContext;
-
-            TransactionTask<T, E> wrappedTask = wrapTaskIfNecessary(task, immutableTsLock);
+        public synchronized void close() {
+            if (hasClosed) {
+                // Some operations in close are not idempotent, e.g. openTransactionCounter.dec().
+                // Let's guarantee we run close() only once, and no-op in following runs to respect Closeable interface.
+                return;
+            }
 
             ExpectationsAwareTransaction txn = delegate;
-            T result;
             try {
-                txn.onCommitOrAbort(txn::reportExpectationsCollectedData);
-                txn.onCommitOrAbort(callback);
-                result = runTaskThrowOnConflict(wrappedTask, txn);
+                if (txn.isUncommitted()) {
+                    txn.abort();
+                }
             } finally {
-                lockWatchManager.requestTransactionStateRemovalFromCache(getTimestamp());
-                postTaskContext = postTaskTimer.time();
                 openTransactionCounter.dec();
```

Comment:
What if this was ALSO moved somewhere higher? to somewhere around the place where we increment this value?

Reply:
I think that would be the most correct thing